### ISSUE TITLE
Macro resolution regression when a #[test] is named 'panic'

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/rust#53675]
Test case:

```rust
#[test]
fn panic() {
    assert!(true)
}
```

```rust
$ rustc +nightly -V
rustc 1.30.0-nightly (63d66494a 2018-08-23)
$ cargo +nightly test
error: cannot determine resolution for the macro `panic`
 --> src/lib.rs:3:5
  |
3 |     assert!(true)
  |     ^^^^^^^^^^^^^
  |
  = note: import resolution is stuck, try simplifying macro imports

error: aborting due to previous error

error: Could not compile `foo`.
```

CC @petrochenkov

**Comment by SimonSapin** [comment]
FWIW in the non-reduced file https://github.com/servo/rust-mozjs/blob/dd338dae9bb68fa0cf23d4da21afc10e0153b27c/src/panic.rs the error is reported twice, one of them on a statement-macro with a semi-colon:

```rust
error: cannot determine resolution for the macro `panic`
  --> tests/panic.rs:46:9
   |
46 |         panic!()
   |         ^^^^^
   |
   = note: import resolution is stuck, try simplifying macro imports

error: cannot determine resolution for the macro `panic`
  --> tests/panic.rs:37:9
   |
37 |         assert!(!function.is_null());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: import resolution is stuck, try simplifying macro imports

error: aborting due to 2 previous errors

**Comment by petrochenkov** [comment]
IIRC, the second case is reported only because the first error (without semicolon) breaks everything.

This issue is kinda known, see `proc-macro-attributes.rs` in 674a5db1a50e25dd60a8ee6669edee9a366c9fab.

**Comment by pnkfelix** [comment]
visited for triage. After some discussion about severity, decided to tag as P-high.

**Comment by nikomatsakis** [comment]
@petrochenkov what is required to fix this bug? Do we need to land something like #53913?

**Comment by petrochenkov** [comment]
@nikomatsakis 
I don't know, I didn't investigate what happens in this case, but #53913 is unlikely to be related.

**Comment by nikomatsakis** [comment]
@petrochenkov this seems kind of important to figure out, no? Have you had any further thoughts? If not, I'd be up to try and investigate a bit and see if I can figure something out.

**Comment by pnkfelix** [comment]
visited for triage. Curious about an update from @petrochenkov about how to approach this and whether they have time to attack it. Assigning to self for further investigation/management.

**Comment by petrochenkov** [comment]
Status update: still plan to investigate this (probably this week).
The delay is mostly due to prioritization - there were other regressions and urgent future-proofing work to complete before beta.

**Comment by petrochenkov** [comment]
Not sure how exactly, but this is already fixed.
https://github.com/rust-lang/rust/pull/54209 adds tests.
